### PR TITLE
Adding support to wrapper for the Journeys API

### PR DESCRIPTION
**Pull request by taledo**

### CODE REVIEW

File: csrest_journey_emails.php [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+         *     'Results' => array(
+         *         {
+         *             'EmailAddress' => The suppressed email address
+         *             'ListID' => The ID of the list this subscriber comes from
+         *         }
+         *     )
+         * }
+         */
+
+        function get_journey_recipients($since = '', $page_number = NULL, $page_size = NULL, $order_field = NULL,
```
Comment (niawahyuni) [inline csrest_journey_emails.php]: is it posisble to remove `$order_field` reference from all the journey_emails endpoint (and the comment above the method)? these endspoints did not have order_field parameter (only order_direction)

File: csrest_clients.php [hunk 2]
```diff
@@ -448,5 +448,24 @@ function get_primary_contact() {
...
+
+
+
+          /**
+         * Gets a list of journeys for the current client
+         * @access public
+         * @return CS_REST_Wrapper_Result A successful response will be an object of the form
+         * array(
+         *     {
+          *     'ListID' => The id of the list
```
Comment (cmtiml) [inline csrest_clients.php]: Minor mis-Align here

File: csrest_journey_emails.php [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+         *     'OrderDirection' => The order direction
+         *     'PageNumber' => The page number for the result set
+         *     'PageSize' => The page size used
+         *     'RecordsOnThisPage' => The number of records returned
+         *     'TotalNumberOfRecords' => The total number of records available
+         *     'NumberOfPages' => The total number of pages for this collection
+         *     'Results' => array(
+         *         {
+         *             'EmailAddress' =>
+		 *			   'Date' =>
```
Comment (cmtiml) [inline csrest_journey_emails.php]: Are we referring to the API document for the return results?
It looks like there is slight difference between the API doc and the result shown here:
https://www.campaignmonitor.com/api/journeys/#journey-email-opens

File: csrest_journey_emails.php [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+         *     'PageNumber' => The page number for the result set
+         *     'PageSize' => The page size used
+         *     'RecordsOnThisPage' => The number of records returned
+         *     'TotalNumberOfRecords' => The total number of records available
+         *     'NumberOfPages' => The total number of pages for this collection
+         *     'Results' => array(
+         *         {
+					'EmailAddress' =>
+					'Date' =>
+					'URL' =>
```
Comment (cmtiml) [inline csrest_journey_emails.php]: Same diff here: https://www.campaignmonitor.com/api/journeys/#journey-email-clicks

File: csrest_journeys.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+         *       'JourneyID' => 
+         *       'Name' => 
+         *       'TriggerType' => 
+         *       'Status' => 
+         *       'Emails' => array(
+         *           {
+         *               'EmailID' => String
+         *               'Name' => String
+         *               'Bounced' => Int
+         *               'Clicked' => Int
```
Comment (cmtiml) [inline csrest_journeys.php]: Slight diff here: https://www.campaignmonitor.com/api/journeys/#getting-journey-summary

File: samples/journeys/get_journey_bounces.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+require_once '../../csrest_journey_emails.php';
+
+$auth = array(
+    'access_token' => 'your access token',
+    'refresh_token' => 'your refresh token');
+$wrap = new CS_REST_JourneyEmails('Email ID to get bounces for', $auth);
+$result = $wrap->get_journey_bounces('Get bounces since', 1, 50, 'email', 'asc');
+//$result = $wrap->get_journey_bounces(date('Y-m-d', strtotime('-30 days')), page, page size, order field, order direction);
```
Comment (cmtiml) [inline samples/journeys/get_journey_bounces.php]: Thank you very much for putting a sample code, can we put it above the code to indicate the calling format?

File: samples/journeys/get_journey_clicks.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+require_once '../../csrest_journey_emails.php';
+
+$auth = array(
+    'access_token' => 'your access token',
+    'refresh_token' => 'your refresh token');
+$wrap = new CS_REST_JourneyEmails('Email ID to get clicks for', $auth);
+$result = $wrap->get_journey_clicks('Get clicks since', 1, 50, 'email', 'asc');
+//$result = $wrap->get_journey_clicks(date('Y-m-d', strtotime('-30 days')), page, page size, order field, order direction);
```
Comment (cmtiml) [inline samples/journeys/get_journey_clicks.php]: Same here, can we move it above the code?

File: samples/journeys/get_journey_opens.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+require_once '../../csrest_journey_emails.php';
+
+$auth = array(
+    'access_token' => 'your access token',
+    'refresh_token' => 'your refresh token');
+$wrap = new CS_REST_JourneyEmails('Email ID to get opens for', $auth);
+$result = $wrap->get_journey_opens('Get opens since', 1, 50, 'email', 'asc');
+//$result = $wrap->get_journey_opens(date('Y-m-d', strtotime('-30 days')), page, page size, order field, order direction);
```
Comment (cmtiml) [inline samples/journeys/get_journey_opens.php]: Same here, can we move it above the code?

File: samples/journeys/get_journey_recipients.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+require_once '../../csrest_journey_emails.php';
+
+$auth = array(
+    'access_token' => 'your access token',
+    'refresh_token' => 'your refresh token');
+$wrap = new CS_REST_JourneyEmails('Email ID to get recipients for', $auth);
+$result = $wrap->get_journey_recipients('Get recipients since', 1, 50, 'email', 'asc');
+//$result = $wrap->get_journey_recipients(date('Y-m-d', strtotime('-30 days')), page, page size, order field, order direction);
```
Comment (cmtiml) [inline samples/journeys/get_journey_recipients.php]: Same here, can we move it above the code?

File: samples/journeys/get_journey_unsubscribes.php [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+require_once '../../csrest_journey_emails.php';
+
+$auth = array(
+    'access_token' => 'your access token',
+    'refresh_token' => 'your refresh token');
+$wrap = new CS_REST_JourneyEmails('Email ID to get unsubscribes for', $auth);
+$result = $wrap->get_journey_unsubscribes('Get unsubscribes since', 1, 50, 'email', 'asc');
+//$result = $wrap->get_journey_unsubscribes(date('Y-m-d', strtotime('-30 days')), page, page size, order field, order direction);
```
Comment (cmtiml) [inline samples/journeys/get_journey_unsubscribes.php]: Same here, can we move it above the code?

File: tests/csrest_journey_emails_test.php [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        $expected_result = new CS_REST_Wrapper_Result($deserialised, $response_code);
+
+        $this->setup_transport_and_serialisation($transport_result, $call_options,
+            $deserialised, $raw_result, NULL, NULL, $response_code);
+
+        $result = $this->wrapper->get_journey_bounces($since);
+
+        $this->assertIdentical($expected_result, $result);
+    }
+
```
Comment (cmtiml) [inline tests/csrest_journey_emails_test.php]: Can we remove the extra lines here?

File: tests/csrest_journeys_test.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+    function testget_journey_summary() {
+
+        $raw_result = 'journey details';
+        $deserialised = array(1,23,4,5);
+        $call_options = $this->get_call_options(trim($this->journey_base_route, '/').'.json');
+        $this->general_test('get_journey_summary', $call_options, $raw_result, $deserialised);
+        
+    }
+
```
Comment (cmtiml) [inline tests/csrest_journeys_test.php]: Can we remove the extra lines here?

File: tests/csrest_test.php [hunk 13]
```diff
@@ -40,6 +40,8 @@ function set_up_inner() {
         $this->wrapper = new CS_REST_General($this->auth, $this->protocol, $this->log_level,
                                              $this->api_host, $this->mock_log, $this->mock_serialiser,
                                              $this->mock_transport);
+
```
Comment (cmtiml) [inline tests/csrest_test.php]: Do we need the extra lines here?

File: csrest_journeys.php (hunk 5 above)
Comment (niawahyuni) [reply csrest_journeys.php]: this one is meant to be `Clicked` - in the processing of updating the API docs

File: csrest_journey_emails.php (hunk 1 above)
Comment (taledo) [reply csrest_journey_emails.php]: @niawahyuni No problem, this is done!

**Comment by taledo** [comment]
Pull request comments have been actioned @cmtiml @niawahyuni

File: csrest_journeys.php [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+         *       'JourneyID' => The journey id 
+         *       'Name' => The name of the journey
+         *       'TriggerType' => The method in which the journey was triggered
+         *       'Status' => The status of the journey
+         *       'Emails' => array(
+         *           {
+         *               'EmailID' => The ID of the email attached to the journey
+         *               'Name' => The name of the email attached to the journey
+         *               'Bounced' => The number of recipients who bounced
+         *               'Clicks' => The total number of recorded clicks
```
Comment (niawahyuni) [inline csrest_journeys.php]: @taledo this one is meant to be `Clicked` - the API docs was just updated couple of minutes ago.. sorry for the confustion

Comment (taledo) [reply csrest_journeys.php]: This is also now done. I will merge! Thank you

**Pull request merged**